baseplate/lib/experiments/__init__.py

```diff
@@ -55,13 +55,29 @@ def __init__(
         backoff: Optional[float] = None,
     ):
         self._filewatcher = FileWatcher(path, json.load, timeout=timeout, backoff=backoff)
+        self._global_cache: Dict[str, Optional[Experiment]] = {}
         self._event_logger = event_logger
+        self.cfg_mtime = 0.0
 
     def make_object_for_context(self, name: str, span: Span) -> "Experiments":
+        config_data: Dict[str, Dict[str, str]] = {}
+        try:
+            config_data, mtime = self._filewatcher.get_data_and_mtime()
+
+            if mtime > self.cfg_mtime:
+                self.cfg_mtime = mtime
+                if self._global_cache:
+                    self._global_cache = {}
```

Comment:
```suggestion
                self._global_cache = {}
```

I think we want to replace the cache regardless of if it's currently empty. 

1. Request A comes in. It tries to fetch an experiment that doesn't exist. This triggers reading the experiments config into memory but doesn't put anything in the cache.
2. Request A blocks on some IO like a database call and yields to the event loop.
2. Request B comes in. It tries to do some bucketing and finds the experiment config has changed. It re-reads from disk but because the global cache is currently empty it doesn't make a new cache object. Both request A and B are now pointing at the same cache despite having different config data. 
3. Request B now blocks on IO and yields.
4. Request A wakes back up and tries to do some more bucketing. This time it buckets for an experiment that _does_ exist and it places the data it calculates into the global cache. 
5. Request A finishes up.
6. Request B wakes back up and tries to use that same experiment. It's in the cache! But oh no, it's based on the old file. We've poisoned the cache with stale data.

Replacing the cache unconditionally is probably safest.

Reply:
Yea That is true. for empty global cache with file been updated between 2 req fetch files.
A fetch gets GC_A, file updated, B fetch from file with GC is empty, gets GC_A.then A parse and insert into old experiment config. GC will keep the old experiment of A.
the reason i do this coz there would be another case i thought about (maybe this case has less side effect)
A reads the data from file, check mtime > 0, yield
B reads the data from file,  check mtime > 0, yield
A gets GC_A {}
if we always update GC
B gets GC_B {}
then might end up some req always gets different version of GC update to its own GC, 

how can we handle this case.

Reply:
When we're running in gevent, rather than using real threads, we don't yield except at points of blocking IO.  So we should be safe from yielding happening anywhere between `get_data_and_mtime()` returning and `make_object_for_context()` returning. If we always replace the cache when the file has changed, we're sure we never have stale data in the cache. The worst case with real threading is that we overzealously invalidate the cache which just means a bit of extra CPU work but not any invalid data. I think that sounds fair?

Reply:
That makes a lot of sense. yea I totally agree and definitely did not think about this stale cache issue before.
Great catch and cannot say more thank you for this guidance and help here.
Learned a  lot from this. Really appreciate.